miden-lib/asm/kernels/transaction/main.masm

```diff
@@ -36,152 +36,94 @@ const.EPILOGUE_END=131081
-    # instructions - see: https://github.com/0xPolygonMiden/miden-vm/issues/1122
-    # emit trace to signal that the execution of the prologue has started
-    push.0 drop
+    push.0 drop                         # TODO: remove line, see miden-vm/#1122
     trace.PROLOGUE_START
 
-    # execute the transaction prologue
     exec.prologue::prepare_transaction
     # => []
 
-    # TODO: we execute `push.0 drop` before `trace` as decorators are not supported without other
-    # instructions - see: https://github.com/0xPolygonMiden/miden-vm/issues/1122
-    # emit trace to signal that the execution of the prologue has ended
-    push.0 drop
+    push.0 drop                         # TODO: remove line, see miden-vm/#1122
     trace.PROLOGUE_END
 
     # Note Processing
     # ---------------------------------------------------------------------------------------------
 
-    # TODO: we execute `push.0 drop` before `trace` as decorators are not supported without other
-    # instructions - see: https://github.com/0xPolygonMiden/miden-vm/issues/1122
-    # emit trace to signal that the notes processing has started
-    push.0 drop
+    push.0 drop                         # TODO: remove line, see miden-vm/#1122
     trace.NOTES_PROCESSING_START
 
-    # get the total number of consumed notes
     exec.memory::get_total_num_consumed_notes
     # => [num_consumed_notes]
 
-    # compute the pointer to the consumed note after the last consumed note (i.e. the pointer at
-    # which the looping should terminate)
+    # compute the memory location after all input notes, i.e. the exit condition
     dup exec.memory::get_consumed_note_ptr loc_store.0
     # => [num_consumed_notes]
 
-    # check if we have any notes to consume
     eq.0 not
     # => [should_loop]
 
-    # loop while we have notes to consume
     while.true
-        # TODO: we execute `push.0 drop` before `trace` as decorators are not supported without other
-        # instructions - see: https://github.com/0xPolygonMiden/miden-vm/issues/1122
-        # emit trace to signal that the note execution has started
-        push.0 drop
+        push.0 drop                     # TODO: remove line, see miden-vm/#1122
         trace.NOTE_EXECUTION_START
+        # => []
 
-        # execute the note setup script
         exec.note::prepare_note
         # => [NOTE_SCRIPT_HASH, NOTE_ARGS]
 
-        # invoke the note script using the dyncall instruction
+        # run note's script
         dyncall
-        # => [OUTPUT_3, OUTPUT_2, OUTPUT_1, OUTPUT_0]
+        # => [X, X, X, X]
 
-        # clean up note script outputs
+        # Clear the stack, the note can leave up to 4 words on the stack due to the dyncall
         dropw dropw dropw dropw
         # => []
 
-        # check if we have more notes to consume and should loop again
         exec.note::increment_current_consumed_note_ptr
-        loc_load.0
-        neq
+        # => [current_consumed_note_ptr]
```

Comment:
Note: I think these APIs should change to pass an explict `idx`, it would make testing possible.